common/rulesys.cpp

```diff
@@ -24,462 +24,451 @@
+	if (!category_name.empty()) {
+		category_type = FindCategory(category_name);
+		if (category_type == InvalidCategory) {
+			LogRulesDetail("Unable to find category '{}'.", category_name);
+			return false;
 		}
 	}
-	int i;
-	int rule_count = CountRules();
-	for (i = 0; i < rule_count; i++) {
-		const RuleInfo &rule = s_RuleInfo[i];
-		if (catname == nullptr || cat == rule.category) {
-			into.push_back(rule.name);
+
+	for (int i = 0; i < CountRules(); i++) {
+		const auto& r = s_RuleInfo[i];
+		if (category_name.empty() || category_type == r.category) {
+			l.push_back(r.name);
 		}
 	}
-	return(true);
+
+	return true;
 }
 
-bool RuleManager::ListCategories(std::vector<const char *> &into) {
-	int i;
-	for (i = 0; i < _CatCount; i++) {
-		into.push_back(s_categoryNames[i]);
+bool RuleManager::ListCategories(std::vector <std::string> &l)
+{
+	for (int i = 0; i < _CatCount; i++) {
+		l.push_back(s_categoryNames[i]);
 	}
-	return(true);
+
+	return true;
 }
 
-bool RuleManager::GetRule(const char *rule_name, std::string &return_value) {
+bool RuleManager::GetRule(std::string rule_name, std::string &rule_value)
+{
 	RuleType type;
 	uint16 index;
-	if (!_FindRule(rule_name, type, index))
+	if (!_FindRule(rule_name, type, index)) {
 		return false;
+	}
 
-	char tmp[255] = "";
-	switch(type) {
+	switch (type) {
 		case IntRule:
-			sprintf(tmp, "%i", m_RuleIntValues[index]);
+			rule_value = fmt::format("{}", m_RuleIntValues[index]);
 			break;
 		case RealRule:
-			sprintf(tmp, "%f", m_RuleRealValues[index]);
+			rule_value = fmt::format("{}", m_RuleRealValues[index]);
 			break;
 		case BoolRule:
-			std::string tmp_val = m_RuleBoolValues[index] ? "true" : "false";
-			sprintf(tmp, "%s", tmp_val.c_str());
+			rule_value = fmt::format("{}", m_RuleBoolValues[index] ? "true" : "false");
 			break;
 	}
 
-	return_value = tmp;
-
 	return true;
 }
 
-bool RuleManager::SetRule(const char *rule_name, const char *rule_value, Database *database, bool db_save, bool reload) {
-	if(rule_name == nullptr || rule_value == nullptr)
-		return(false);
+bool RuleManager::SetRule(std::string rule_name, std::string rule_value, Database *db, bool db_save, bool reload)
+{
+	if (rule_name.empty() || rule_value.empty()) {
+		return false;
+	}
 
 	RuleType type;
 	uint16 index;
-	if(!_FindRule(rule_name, type, index))
-		return(false);
+	if (!_FindRule(rule_name, type, index)) {
+		return (false);
+	}
 
 	if (reload) {
-		if (strcasecmp(rule_name, "World:ExpansionSettings") == 0)
-			return(false);
-		if (strcasecmp(rule_name, "World:UseClientBasedExpansionSettings") == 0)
-			return(false);
+		bool is_client = Strings::Contains(rule_name, "World:UseClientBasedExpansionSettings");
+		bool is_world = Strings::Contains(rule_name, "World:ExpansionSettings");
+		if (is_client || is_world) {
+			return false;
+		}
 	}
 
-	switch(type) {
+	switch (type) {
 		case IntRule:
-			m_RuleIntValues[index] = atoi(rule_value);
+			m_RuleIntValues[index] = std::stoi(rule_value);
```

Comment:
I'd recommend using std::strtol if changing these since it doesn't throw. `std::stoi` and variants are just wrappers that throw exceptions instead of returning 0 for bad input. Either that or catch the exceptions and log bad inputs since changing this is a behavior change and can throw now